core/server/master/src/main/java/alluxio/master/file/AccessTimeUpdater.java

```diff
@@ -0,0 +1,176 @@
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This class handles the update of inode last access time.
+ */
+@NotThreadSafe
+final class AccessTimeUpdater implements JournalSink {
+  private static final Logger LOG = LoggerFactory.getLogger(AccessTimeUpdater.class);
+
+  private final long mFlushInterval;
+  private final long mUpdatePrecision;
+  private final long mShutdownTimeout;
+
+  private final FileSystemMaster mFileSystemMaster;
+  private final InodeTree mInodeTree;
+
+  /** Keep track of all inodes that need access time update. */
+  private ConcurrentHashSet<Long> mAccessTimeUpdates;
+  private ScheduledExecutorService mExecutorService = null;
+  private AtomicBoolean mUpdateScheduled = new AtomicBoolean();
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater}.
+   */
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem) {
+    this(fileSystemMaster, inodeTree, journalSystem,
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_JOURNAL_FLUSH_INTERVAL),
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATE_PRECISION),
+        ServerConfiguration.getMs(PropertyKey.MASTER_FILE_ACCESS_TIME_UPDATER_SHUTDOWN_TIMEOUT));
+  }
+
+  /**
+   * Constructs a new {@link AccessTimeUpdater} with time configurations.
+   */
+  @VisibleForTesting
+  public AccessTimeUpdater(FileSystemMaster fileSystemMaster, InodeTree inodeTree,
+      JournalSystem journalSystem, long flushInterval, long updatePrecision, long shutdownTimeout) {
+    mFileSystemMaster = fileSystemMaster;
+    mInodeTree = inodeTree;
+    mAccessTimeUpdates = new ConcurrentHashSet<>();
+    mFlushInterval = flushInterval;
+    mUpdatePrecision = updatePrecision;
+    mShutdownTimeout = shutdownTimeout;
+    journalSystem.addJournalSink(mFileSystemMaster, this);
+  }
+
+  public void start() {
+    start(mFlushInterval > 0 ? Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("AccessTimeUpdater-%d", true)) : null);
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executorService) {
+    mExecutorService = executorService;
+  }
+
+  @Override
+  public void beforeShutdown() {
+    if (mExecutorService != null) {
+      flushUpdates();
+    }
+  }
+
+  public void stop() {
+    if (mExecutorService != null) {
+      ThreadUtils.shutdownAndAwaitTermination(mExecutorService, mShutdownTimeout);
+    }
+  }
+
+  /**
+   * Update inode last access time. Requires at least read lock acquired for the inode.
+   *
+   * @param context the journal context
+   * @param inode the target inode
+   * @param opTimeMs the last access time to set on the inode
+   */
+  public void updateAccessTime(JournalContext context, Inode inode, long opTimeMs) {
+    if (opTimeMs - inode.getLastAccessTimeMs() > mUpdatePrecision) {
+      try (LockResource lr = mInodeTree.getInodeLockManager().lockUpdate(inode.getId())) {
```

Comment:
This is to acquire update lock which is only for updating a timestamp. This needs to be done in addition to the expected read lock.

Reply:
How is this possible? We are holding a read lock, and then we are trying to write lock it? Lock upgrades is not possible with java locks.

Also, is it required to lock again? If we only have the read lock, we could have out-of-order access time entries, but I thought that was fine, since we handle that on the replay side?

Reply:
Looking at the implementation of `lockUpdate()`, it basically locks a different type of lock instead of the inode lock. Here is the explanation from the javadoc:

> Locks for guarding changes to last modified time and size on read-locked parent inodes.
> 
> When renaming, creating, or deleting, we update the last modified time and size of the parent inode while holding only a read lock. In the presence of concurrent operations, this could cause the last modified time to decrease, or lead to incorrect directory sizes. To avoid this, we guard the parent inode read-modify-write with this lock. To avoid deadlock, a thread should never acquire more than one of these locks at the same time, and no other locks should be taken while holding one of these locks.